### PR TITLE
22.3 fb request form approved notification

### PR DESCRIPTION
**Pull request by Ohsudev**
Modified program to implement a process to generate email notification when a request is approved.

### CODE REVIEW

File: onprc_ehr/resources/scripts/onprc_ehr/onprc_triggers.js [hunk 1]
```diff
@@ -1165,7 +1165,110 @@ exports.init = function(EHR){
...
+    EHR.Server.TriggerManager.registerHandlerForQuery(EHR.Server.TriggerManager.Events.BEFORE_UPSERT, 'ehr', 'requests', function (helper, scriptErrors, row, oldRow) {
+        if (row.requestid && row.qcstate.label ==  'Request: Approved'){
+
+                var requestid = [];
+                for (var i in row.requestid.length){
+                    var rows = row.requestid[i];
+                    requestid.push(i);
+                }
+
+        }
```
Comment (labkey-martyp) [inline onprc_ehr/resources/scripts/onprc_ehr/onprc_triggers.js]: You don't need this

File: onprc_ehr/resources/scripts/onprc_ehr/onprc_triggers.js [hunk 2]
```diff
@@ -1165,7 +1165,110 @@ exports.init = function(EHR){
...
+
+            var msgs =   helper.getJavaHelper().sendRequestStateEmail("Request: Approved", row.requestid );
+            if (msgs && msgs.length){
+                LABKEY.ExtAdapter.each(msgs, function(msg){
+                    EHR.Server.Utils.addError(scriptErrors, 'qcstate', msg, 'INFO');
+                }, this);
+            }
+
+        }
+    });
```
Comment (labkey-martyp) [inline onprc_ehr/resources/scripts/onprc_ehr/onprc_triggers.js]: You have two of these AFTER_UPSERT triggers for requests. Looks like this one might be the duplicate that can be deleted.

File: onprc_ehr/resources/scripts/onprc_ehr/onprc_triggers.js [hunk 3]
```diff
@@ -1165,7 +1165,110 @@ exports.init = function(EHR){
...
+                    requestid.push(i);
+                }
+
+        }
     });
 
+    //Added: 10-4-2022  R.Blasa
+    EHR.Server.TriggerManager.registerHandlerForQuery(EHR.Server.TriggerManager.Events.AFTER_UPSERT, 'ehr', 'requests', function(helper, errors, row, oldRow){
+        if (row.requestid && row.qcstate.label ==  'Request: Approved'){
+            var requestsApproved = triggerHelper.getApprovedArray();
```
Comment (labkey-martyp) [inline onprc_ehr/resources/scripts/onprc_ehr/onprc_triggers.js]: I don't see getApprovedArray function in this PR. But I don't think you need to do any extra handling here for batch updates. row.requestid should always be a single value.

File: onprc_ehr/resources/scripts/onprc_ehr/onprc_triggers.js [hunk 4]
```diff
@@ -1165,7 +1165,110 @@ exports.init = function(EHR){
...
+            var msgs =   helper.getJavaHelper().sendRequestStateEmail("Request: Approved", requestsApproved );
```
Comment (labkey-martyp) [inline onprc_ehr/resources/scripts/onprc_ehr/onprc_triggers.js]:
```suggestion
            var msgs =   helper.getJavaHelper().sendRequestStateEmail("Request: Approved", [row.requestid] );
```
row.requestid should always be a single value in this handler so just making a single value array out of it should work fine.

**Review by labkey-martyp — COMMENTED** [review]
Hi Raymond, I added some comments. Since it doesn't look like you'll be needing to support batch updates, I think you can simplify the trigger. I did not test this PR but added some suggestions that I think are all you need for this.

**Comment by labkey-martyp** [comment]
Ok Raymond, I added the approved request tracking to the EHR as it may be helpful for others and made the TriggerScriptHelper.sendRequestStateEmail public for use here. That [PR is here](https://github.com/LabKey/ehrModules/pull/424), I will try to get that merged today.

I'll put a patch in the ticket (it won't let me attach it here).  Once the EHR PR is merged in, you can apply the patch to this fb and it will use that new EHR functionality and send notifications (batched if necessary) on request approval.

**Review by kollil — APPROVED** [review]

**Pull request merged**